### PR TITLE
cose sign

### PR DESCRIPTION
**Pull request by franziskuskiefer**
There are still many things hard coded in here (like the algorithm and thus key sizes etc.). But it demonstrates that all parts work together.

We don't actually need signing later. But to generate test cases this will probably be handy.

### CODE REVIEW

File: src/cose/cose.rs [hunk 1]
```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
...
 
-/// Sign the payload and return a serialised COSE_Sign object.
-#[allow(unused_variables)]
-pub fn sign(payload: &[u8]) -> Result<Vec<u8>, CoseError> {
-    unimplemented!()
+// XXX: This works only with P256 for now!
+fn build_pkcs8(public_key: &[u8], secret_key: &[u8]) -> Vec<u8> {
+    let mut pkcs8: Vec<u8> =
+        vec![
+      0x30, 0x81, 0x87, 0x02, 0x01, 0x00, 0x30, 0x13, 0x06, 0x07, 0x2a, 0x86,
```
Comment (mozkeeler) [inline src/cose/cose.rs]: We should probably document what these magic bytes are.

File: src/cose/cose.rs [hunk 2]
```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
...
+fn build_spki(key: &[u8]) -> Vec<u8> {
+    let mut spki = vec![0x30, 0x59, 0x30, 0x13, 0x06, 0x07, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x02,
+                        0x01, 0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x03, 0x01, 0x07, 0x03,
+                        0x42, 0x00];
+    spki.extend_from_slice(key);
+    spki
+}
+
+/// The key has to be an uncrompressed curve point on the point specified in alg.
+pub fn sign(
```
Comment (mozkeeler) [inline src/cose/cose.rs]: This seems like an API we're really only implementing for the purposes of testing the verification side of the API. Should this just go in some internal test module?

File: src/cose/cose.rs [hunk 3]
```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
...
+    let signature = signature.unwrap();
+
+    let mut cose_signature: Vec<CborType> = Vec::new();
+    // TODO: add intermediate certificates to the protected header.
+    cose_signature.push(CborType::Bytes(vec![]));
+
+    // No body headers
+    cose_signature.push(CborType::Map(BTreeMap::new()));
+
+    // The payload is empty (TODO: it should be nil)
```
Comment (mozkeeler) [inline src/cose/cose.rs]: File an issue on supporting `nil`?

File: src/cose/cose.rs [hunk 4]
```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
...
+    let mut sig_structure: Vec<CborType> = Vec::new();
+    sig_structure.push(CborType::String("Signature".to_string()));
+    // TODO: add the protected body header here.
+    sig_structure.push(CborType::Bytes(vec![]));
+
+    // Encode values into the protected_sig_header, which is a byte array
+    // encoding a map.
+    if alg != CoseSignatureType::ES256 {
+        // TODO: we only accept ES256 for now.
+        return Vec::new();
```
Comment (mozkeeler) [inline src/cose/cose.rs]: `return Err(CoseError::UnknownSignatureScheme);`? (or maybe `Unsupported` or `Unimplemented`?)

File: src/cose/cose.rs [hunk 5]
```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
...
+    }
+    let mut header_map: BTreeMap<CborType, CborType> = BTreeMap::new();
+    header_map.insert(CborType::Integer(1), CborType::SignedInteger(-7));
+    let header_map = CborType::Map(header_map).serialize();
+    sig_structure.push(CborType::Bytes(header_map));
+
+    // XXX: We don't hande external_aad here.
```
Comment (mozkeeler) [inline src/cose/cose.rs]: Let's not have "XXX" comments. If it's a todo we're actually going to do, let's have an issue on file for it and reference it here. If it's not something we're ever going to support, we can just say "This doesn't support \<thing\>" either in the code or in the API docs.

File: src/cose/cose.rs [hunk 6]
```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
...
+    sig_structure.push(CborType::Bytes(vec![]));
+
+    sig_structure.push(CborType::Bytes(payload.to_vec()));
+
+    let sig_structure = CborType::Array(sig_structure);
+    sig_structure.serialize()
+}
+
+fn build_spki(key: &[u8]) -> Vec<u8> {
+    let mut spki = vec![0x30, 0x59, 0x30, 0x13, 0x06, 0x07, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x02,
```
Comment (mozkeeler) [inline src/cose/cose.rs]: Document magic bytes.

File: src/cose/cose.rs [hunk 7]
```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
...
+      0x48, 0xce, 0x3d, 0x02, 0x01, 0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d,
+      0x03, 0x01, 0x07, 0x04, 0x6d, 0x30, 0x6b, 0x02, 0x01, 0x01, 0x04, 0x20,
+      // Private key: len, bytes.
+      // Public key (uncompressed EC point for now): 0xa1, len, type, len, 0x00, bytes.
+      ];
+    pkcs8.extend_from_slice(secret_key);
+    pkcs8.push(0xa1);
+    let len = (public_key.len() + 4 - 1) as u8;
+    pkcs8.push(len);
+    pkcs8.push(0x03);
```
Comment (mozkeeler) [inline src/cose/cose.rs]: // BIT STRING

File: src/cose/cose.rs [hunk 8]
```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
...
+    pkcs8.push(len - 2);
+    pkcs8.push(0x00);
```
Comment (mozkeeler) [inline src/cose/cose.rs]: // zero unused bits

File: src/cose/cose.rs [hunk 9]
```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
...
+
+/// The key has to be an uncrompressed curve point on the point specified in alg.
+pub fn sign(
+    payload: &[u8],
+    alg: CoseSignatureType,
+    public_key: &[u8],
+    secret_key: &[u8],
+) -> Result<Vec<u8>, CoseError> {
+    let nss_alg = match alg {
+        CoseSignatureType::ES256 => nss::SignatureAlgorithm::ES256,
```
Comment (mozkeeler) [inline src/cose/cose.rs]: Seems a bit unnecessary to have multiple copies of these algorithms - let's unify them in a follow-up patch.

File: src/cose/cose.rs [hunk 10]
```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
...
+    sig_structure.push(CborType::Bytes(vec![]));
+
+    // Encode values into the protected_sig_header, which is a byte array
+    // encoding a map.
+    if alg != CoseSignatureType::ES256 {
+        // TODO: we only accept ES256 for now.
+        return Vec::new();
+    }
+    let mut header_map: BTreeMap<CborType, CborType> = BTreeMap::new();
+    header_map.insert(CborType::Integer(1), CborType::SignedInteger(-7));
```
Comment (mozkeeler) [inline src/cose/cose.rs]: We want the key id header too, right?

File: src/cose/cose.rs [hunk 11]
```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
...
+///        context : "Signature" / "Signature1" / "CounterSignature",
+///        body_protected : empty_or_serialized_map,
+///        ? sign_protected : empty_or_serialized_map,
+///        external_aad : bstr,
+///        payload : bstr
+///    ]
+fn build_signature_structure(payload: &[u8], alg: CoseSignatureType) -> Vec<u8> {
+    let mut sig_structure: Vec<CborType> = Vec::new();
+    sig_structure.push(CborType::String("Signature".to_string()));
+    // TODO: add the protected body header here.
```
Comment (mozkeeler) [inline src/cose/cose.rs]: Yeah this seems like something we should do.

File: src/cose/cose.rs [hunk 12]
```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
...
+) -> Result<Vec<u8>, CoseError> {
+    let nss_alg = match alg {
+        CoseSignatureType::ES256 => nss::SignatureAlgorithm::ES256,
+        _ => return Err(CoseError::UnkownSignatureScheme),
+    };
+    let pkcs8 = build_pkcs8(public_key, secret_key);
+
+    // The protected signature header contains
+    //  * the signature algorithm
+    //  * XXX: the DER encoded EE certificate (not yet)
```
Comment (mozkeeler) [inline src/cose/cose.rs]: "TODO: some issue #"

File: src/cose/cose.rs (hunk 9 above)
Comment (franziskuskiefer) [reply src/cose/cose.rs]: Filed #14

File: src/cose/cose.rs (hunk 11 above)
Comment (franziskuskiefer) [reply src/cose/cose.rs]: I filed issue #15 to make the signature compliant with the way we defined add-on signatures.

**Comment by franziskuskiefer** [comment]
> In general looks good. My main question is why we need this as a public API in cose - it's just for tests, right?

Right, we don't this as a public API at the moment. I can make it private

>  Also, I realize we're iterating quickly on this, but I'd rather avoid NSS-style things like "XXX fixme later". TODOs are fine, but let's actually file issues and reference them. That way we have a concrete list of work laid out for us.

Agreed.

File: src/cose/cose.rs (hunk 10 above)
Comment (franziskuskiefer) [reply src/cose/cose.rs]: Yeah eventually. right now we don't have it in here yet.
I think I'd like to land a version of this patch and then rewrite both, verify and sign, to use the correct format (#15).

**Comment by franziskuskiefer** [comment]
Abandoning this PR in favour of #30.

**Pull request closed without merging**